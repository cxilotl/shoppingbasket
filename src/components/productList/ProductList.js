import React from 'react';
import PropTypes from 'prop-types';
import cssStyles from "./ProductList.module.scss";
import Product from "../../containers/Product";

const generateProductList = (products) => {
  let productListContent;
  if (products && Array.isArray(products) && products.length > 0) {
    productListContent = (
      <>
        {
          products.map((product, index) => {
            return (
              <li key={ product.name } className={ cssStyles.productListItem } data-testid={ `${product.name}-${index}` }>
                <Product
                  name={ product.name }
                  price={ product.price }
                  isPerUnit={ product.isPricePerUnit }
                  isPerKg={ product.isPricePerKg }
                />
              </li>
            );
          })
        }
      </>
    );
  }
  return productListContent;
};

const ProductList = ({ products }) => {
  const productList = generateProductList(products);
  return (
    <ul className={ cssStyles.productList }>
      { productList }
    </ul>
  );
};

ProductList.propTypes = {
  products:PropTypes.array
};

export default ProductList;