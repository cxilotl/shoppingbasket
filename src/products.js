export default [
  {
    name: 'Beans',
    price: 0.5,
    isPricePerUnit: true,
    isPricePerKg: false
  },
  {
    name: 'Coke',
    price: 0.7,
    isPricePerUnit: true,
    isPricePerKg: false
  },
  {
    name: 'Oranges',
    price: 1.99,
    isPricePerUnit: false,
    isPricePerKg: true
  }
];