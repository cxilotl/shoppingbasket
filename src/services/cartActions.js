import { ADD_TO_CART, REMOVE_FROM_CART } from './actionTypes';

const addProduct = (product) => {
  return {
    type: ADD_TO_CART,
    payload: {
      product: product
    }
  }
};

const removeProduct = (product) => {
  return {
    type: REMOVE_FROM_CART,
    payload: {
      product: product
    }
  }
};

export {
  addProduct,
  removeProduct
};