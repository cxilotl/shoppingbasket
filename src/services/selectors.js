const getProducts = (state) => {
  return state.products;
};

const getProduct = (state, name) => {
  return state.products.find((product) => {
    return product.name === name;
  });
};

const isCartEmpty = (state) => {
  return (state.cart && state.cart.length === 0);
};

const getProductFromCart = (state, name) => {
  return state.cart.find((product) => {
    return product.name === name;
  });
};

const getTotalToPay = (state) => {
  return state.cart.reduce((total, productInCart) => {
    const productPricePerUnit = getProduct(state, productInCart.name);
    const totalPriceOfProduct = productPricePerUnit.price * productInCart.quantity;
    return total + totalPriceOfProduct;
  }, 0);
};

const getCartItems = (state) => {
  return state.cart.map((productInCart) => {
    return Object.assign({}, productInCart, {
      price: getProduct(state, productInCart.name).price
    });
  });
};

export {
  getProducts,
  getProduct,
  isCartEmpty,
  getProductFromCart,
  getTotalToPay,
  getCartItems
};