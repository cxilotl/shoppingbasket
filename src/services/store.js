import { createStore } from "redux";
import reducer from './reducers';
import products from '../products';

const store = createStore(
  reducer,
  {
    products: products,
    cart: []
  },
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__() // for debugging
);

export { store };