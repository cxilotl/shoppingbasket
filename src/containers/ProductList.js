import { connect } from 'react-redux';
import ProductList from "../components/productList/ProductList";
import { getProducts } from "../services/selectors";

const mapStateToProps = (state) => {
  return {
    products: getProducts(state)
  };
};

const mapDispatchToProps = null;

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ProductList);