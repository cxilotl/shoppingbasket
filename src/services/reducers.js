import { ADD_TO_CART, REMOVE_FROM_CART } from './actionTypes';
import { isCartEmpty, getProductFromCart } from "./selectors";

const handleAddProductToCart = (state, payload) => {
  if (typeof payload.product !== 'undefined') {
    if (!isCartEmpty(state)) {
      if (getProductFromCart(state, payload.product)) {
        return {
          ...state,
          cart: state.cart.map((productItem) => {
            if (productItem.name === payload.product) {
              return {
                name: productItem.name,
                quantity: productItem.quantity + 1
              }
            } else {
              return productItem;
            }
          })
        };
      } else {
        return {
          ...state,
          cart: [
            ...state.cart,
            {
              name: payload.product,
              quantity: 1
            }
          ]
        };
      }
    } else {
      return {
        ...state,
        cart: [{
          name: payload.product,
          quantity: 1
        }]
      };
    }
  }
  return state;
};

const handleRemoveProductFromCart = (state, payload) => {
  if (typeof payload.product !== 'undefined' && !isCartEmpty(state) && getProductFromCart(state, payload.product)) {
    const existingProduct = getProductFromCart(state, payload.product);
    if (existingProduct.quantity > 1) {
      return {
        ...state,
        cart: state.cart.map((productItem) => {
          if (productItem.name === payload.product) {
            return {
              name: productItem.name,
              quantity: productItem.quantity - 1
            }
          } else {
            return productItem;
          }
        })
      }
    } else {
      return {
        ...state,
        cart: state.cart.filter((product) => {
          return product.name !== payload.product;
        })
      };
    }
  }
  return state;
};

const initialState = {
  products: [],
  cart: []
};

const reducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case ADD_TO_CART:
      return handleAddProductToCart(state, action.payload);
    case REMOVE_FROM_CART:
      return handleRemoveProductFromCart(state, action.payload);
    default:
      return state;
  }
};

export default reducer;