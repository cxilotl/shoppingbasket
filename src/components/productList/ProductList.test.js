import React from 'react';
import { Provider } from "react-redux";
import renderer from 'react-test-renderer';
import ProductList from "./ProductList";

import { store as appStore } from '../../services/store';
import products from "../../products";

describe('Product List', () => {
  it('Snapshot', () => {
    const component = renderer.create(
      <Provider store={ appStore }>
        <ProductList products={ products }/>
      </Provider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});