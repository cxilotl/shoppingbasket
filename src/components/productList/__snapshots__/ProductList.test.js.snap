// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product List Snapshot 1`] = `
<ul
  className="productList"
>
  <li
    className="productListItem"
    data-testid="Beans-0"
  >
    <div
      className="product"
    >
      <h4
        className="title"
      >
        Beans
      </h4>
      <p
        className="price"
      >
        £0.5
      </p>
      <div
        className="controls"
      >
        <button
          name="product-increment-btn"
          onClick={[Function]}
          type="button"
        >
          +
        </button>
        <button
          name="product-decrement-btn"
          onClick={[Function]}
          type="button"
        >
          -
        </button>
      </div>
    </div>
  </li>
  <li
    className="productListItem"
    data-testid="Coke-1"
  >
    <div
      className="product"
    >
      <h4
        className="title"
      >
        Coke
      </h4>
      <p
        className="price"
      >
        £0.7
      </p>
      <div
        className="controls"
      >
        <button
          name="product-increment-btn"
          onClick={[Function]}
          type="button"
        >
          +
        </button>
        <button
          name="product-decrement-btn"
          onClick={[Function]}
          type="button"
        >
          -
        </button>
      </div>
    </div>
  </li>
  <li
    className="productListItem"
    data-testid="Oranges-2"
  >
    <div
      className="product"
    >
      <h4
        className="title"
      >
        Oranges
      </h4>
      <p
        className="price"
      >
        £1.99/kg
      </p>
      <div
        className="controls"
      >
        <button
          name="product-increment-btn"
          onClick={[Function]}
          type="button"
        >
          +
        </button>
        <button
          name="product-decrement-btn"
          onClick={[Function]}
          type="button"
        >
          -
        </button>
      </div>
    </div>
  </li>
</ul>
`;
